### PR TITLE
Parallel_restore format v2 support

### PR DESCRIPTION
**Pull request by h-shaffer**
Add support for new format v2 structures. In this branch only scoutfs filesystem format v2 will be supported

### CODE REVIEW

**Comment by versity-github** [comment]
This won't automatically run in continuous integration without approval. A member of the Versity organization must allow it.

**Review by aversecat — CHANGES_REQUESTED** [review]
You wrote a few commit messages that are all in the title (first line) and thus overflowed into the body of the commit message.

You should always have a one-line (sub-72 character) title. Then an empty line, and then the elaborate commit message with s-o-b.

File: tests/tests/parallel_restore.sh
Comment (aversecat) [inline tests/tests/parallel_restore.sh:17]: I have a separate commit to fix this bug - we should split it out because we can squash it away before it gets merged into `main` eventually.

File: utils/src/parallel_restore.c
Comment (aversecat) [inline utils/src/parallel_restore.c:1911]: Generally you shouldn't do this and the NULL assignment right before the allocation. Since read_block allocates, you want to start with blk = NULL (and you properly initialize it above), but because you free(blk) before read_block() is called, blk will only be freed if there are more iterations in the loop. On the last iteration, blk is never freed and you leak the memory.

In other words, just do read_block() ... free() every loop in that order, and make sure to account for error cases.

**Review by aversecat — CHANGES_REQUESTED** [review]
You have a title here and the body is indented - probably because there is an empty line missing between them.

**Review by aversecat — CHANGES_REQUESTED** [review]
Missing s-o-b here too...

File: utils/src/parallel_restore.h
Comment (aversecat) [inline utils/src/parallel_restore.h:49]: I wouldn't do these #ifdefs here. These flags are universal for format V2 and they are going to be needed for all sorts of other flags in the future - they aren't tied to retention alone. Just like with the project ID we should always generate v2 content on disk when restoring.

So please just remove the ifdefs from this and the project ID patch, and if someone makes a wild version where it doesn't compile, they're on their own.

File: utils/src/parallel_restore.c
Comment (aversecat) [inline utils/src/parallel_restore.c:408]: drop the ifdefs here, too

File: utils/src/parallel_restore.c
Comment (aversecat) [inline utils/src/parallel_restore.c:930]: drop this one too.

File: tests/src/parallel_restore.c
Comment (aversecat) [inline tests/src/parallel_restore.c:264]: Maybe call it TEST_RULE_STR instead?

File: tests/src/parallel_restore.c
Comment (aversecat) [inline tests/src/parallel_restore.c:272]: space after comma

File: tests/src/parallel_restore.c
Comment (aversecat) [inline tests/src/parallel_restore.c:626]: probably an editor accident

File: tests/src/parallel_restore.c
Comment (aversecat) [inline tests/src/parallel_restore.c:267]: the function here generates a single quota rule. The name should reflect this.

File: tests/golden/parallel_restore
Comment (aversecat) [inline tests/golden/parallel_restore:19]: You should verify that we can actually see that the quota rule was properly added by listing the quota rules for the filesystem here.

File: tests/src/parallel_restore.c
Comment (aversecat) [inline tests/src/parallel_restore.c:635]: Another thing to add is to implement quota restore in `restore_copy.c`.

**Review by aversecat — CHANGES_REQUESTED** [review]

File: tests/tests/parallel_restore.sh
Comment (h-shaffer) [reply tests/tests/parallel_restore.sh:17]: Hmm, github isn't making it too clear what you're referencing here, could you elaborate a bit?

File: utils/src/parallel_restore.h
Comment (h-shaffer) [reply utils/src/parallel_restore.h:49]: The original thought for this was that these constants were guaranteed to exist on a format v2 fs where they wouldn't on fmt < v2. However, I agree they should be removed now that we only allow version v2.

**Comment by aversecat** [comment]
You are deleting these lines because a different test shows that they were
added by mistake. The problem is that the removal should be a separate
commit explaining why they need to be removed.

I have a patch for this already. You can redo yours and make it into a
separate commit, or just drop it.

File: tests/src/parallel_restore.c
Comment (h-shaffer) [reply tests/src/parallel_restore.c:635]: Cleaned up everything else, I'm going to push those changes up while I work on this

File: utils/src/parallel_restore.h
Comment (aversecat) [inline utils/src/parallel_restore.h:80]: This should be a different length padding, don't you agree? Moreover, `names` isn't aligned at all right now, so we'd want another pad in front to make that align. But I really wonder if you just can't use `struct scoutfs_ioctl_quota_rule` instead, which has all the correct padding.

File: utils/src/parallel_restore.c
Comment (aversecat) [inline utils/src/parallel_restore.c:1902]: Don't fix up unrelated bugs without any comments like this. Make them a separate commit.

File: tests/tests/parallel_restore.sh
Comment (aversecat) [inline tests/tests/parallel_restore.sh:67]: This is duplicating a bunch of existing tests. merge issue?

File: utils/src/parallel_restore.c
Comment (aversecat) [inline utils/src/parallel_restore.c]: still has the semicolon here... please remove it.

or just write `if (!blk)`

File: utils/src/parallel_restore.c
Comment (aversecat) [inline utils/src/parallel_restore.c:1905]: You can't free it here yet. You're reading its contents 6 lines below.

File: utils/src/parallel_restore.c
Comment (aversecat) [inline utils/src/parallel_restore.c:1970]: remove this.

File: tests/src/parallel_restore.c
Comment (aversecat) [inline tests/src/parallel_restore.c:634]: `generate_quota` allocates memory, and returns a pointer to it. The caller should then free this memory after it's been used.

**Review by aversecat — CHANGES_REQUESTED** [review]

File: utils/src/parallel_restore.c
Comment (h-shaffer) [reply utils/src/parallel_restore.c]: Hmm I was pretty certain that this got fixed in my changes but it doesn't seem so, I'll make sure its fixed in the next push

File: utils/src/parallel_restore.c
Comment (aversecat) [inline utils/src/parallel_restore.c:1837]: No, I think this is bad API design. The library should stay consistent and not free pointers here. The data here could have been on the stack, or part of some other struct that the caller maintains, like an array.

The caller should free the `rule` after `scoutfs_parallel_restore_add_quota_rule()` returns. Ths same way that `scoutfs_parallel_restore_add_entry` and `scoutfs_parallel_restore_add_inode` also do not free the pointer.

In general, if the library allocates the memory, it should by default also free() it when it is appropriate. For stuff that's just passed in by pointer, it should by default not free the memory.

File: utils/src/parallel_restore.c
Comment (aversecat) [inline utils/src/parallel_restore.c:822]: Can we have a quick test case to make sure this works?

File: utils/src/parallel_restore.c
Comment (aversecat) [inline utils/src/parallel_restore.c:831]: Same here, we should have a test that validates that this works.

**Review by aversecat — CHANGES_REQUESTED** [review]
Just one change and some extra tests needed I think now.

File: tests/tests/parallel_restore.sh
Comment (aversecat) [inline tests/tests/parallel_restore.sh:71]: The test output of these things ended up in the next patch by accident.

**Comment by aversecat** [comment]
Obsolete now, but we should make sure we revisit the outstanding items against the other branch.

**Pull request closed without merging**
